File: simulate_scoring.py

import re

def normalize_text(value):
    s = str(value or '').strip().lower()
    if not s: return ''
    # Simple normalization for python
    s = s.replace("’", "'")
    s = re.sub(r'\s+', ' ', s)
    return s.strip()

def has_word(needle, msg_norm):
    n = str(needle or '').strip()
    if not n: return False
    esc = re.escape(n)
    try:
        return re.search(r'\b' + esc + r'\b', msg_norm, re.IGNORECASE) is not None
    except:
        return False

def score_title(title, user_message):
    msg_norm = normalize_text(user_message)
    t = normalize_text(title)
    stop = set(['della','delle','degli','dello','dell','d','del','dei','di','da','a','al','allo','alla','alle','ai','il','lo','la','i','gli','le','un','uno','una','and','or','the','of','to','in','on','for'])

    score = 0
    # 1) Exact match or full substring (high confidence)
    if msg_norm == t:
        score += 15
    elif t in msg_norm:
        score += 12
    elif msg_norm in t and len(msg_norm) >= 4:
        score += 10

    # 2) Token matching (word boundaries)
    tokens = [tok for tok in t.split(' ') if len(tok) >= 3 and tok not in stop]
    hits = 0
    for tok in tokens:
        if has_word(tok, msg_norm) or (len(msg_norm) >= 4 and msg_norm in tok) or (len(tok) >= 4 and tok in msg_norm):
            hits += 1
            score += 4
            print(f"  Match token: {tok}")

    # 3) Bonus for matching more tokens
    if hits > 0 and hits == len(tokens):
        score += 5

    # 4) If user message is exactly one of the tokens
    if len(tokens) > 1 and any(tok == msg_norm for tok in tokens):
        score += 8

    return score

titles = [
    "Smoothie Giallo Passion",
    "Smoothie Rosso Berry",
    "Smoothie Verde Boost"
]

query = "smoothie berry"

print(f'Query: "{query}"')
for title in titles:
    print(f'Title: "{title}"')
    score = score_title(title, query)
    print(f'Score: {score}')
    print('---')

File: simulate_scoring.js

const normalizeText = (value) => {
    const s = String(value ?? '').trim().toLowerCase();
    if (!s) return '';
    try {
      return s
        .normalize('NFD')
        .replace(/[\u0300-\u036f]/g, '')
        .replace(/[’']/g, "'")
        .replace(/\s+/g, ' ')
        .trim();
    } catch {
      return s.replace(/\s+/g, ' ').trim();
    }
};

const hasWord = (needle, msgNorm) => {
    const n = String(needle || '').trim();
    if (!n) return false;
    const esc = n.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    try { return new RegExp(`\\b${esc}\\b`, 'i').test(msgNorm); } catch { return false; }
};

const scoreTitle = (title, userMessage) => {
    const msgNorm = normalizeText(userMessage);
    const t = normalizeText(title);
    const stop = new Set(['della','delle','degli','dello','dell','d','del','dei','di','da','a','al','allo','alla','alle','ai','il','lo','la','i','gli','le','un','uno','una','and','or','the','of','to','in','on','for']);

    let score = 0;
    // 1) Exact match or full substring (high confidence)
    if (msgNorm === t) {
        score += 15;
    } else if (msgNorm.includes(t)) {
        score += 12;
    } else if (t.includes(msgNorm) && msgNorm.length >= 4) {
        score += 10;
    }

    // 2) Token matching (word boundaries)
    const tokens = t.split(' ').filter(tok => tok.length >= 3 && !stop.has(tok));
    let hits = 0;
    for (const tok of tokens) {
        if (hasWord(tok, msgNorm) || (msgNorm.length >= 4 && tok.includes(msgNorm)) || (tok.length >= 4 && msgNorm.includes(tok))) {
            hits++;
            score += 4;
            console.log(`  Match token: ${tok}`);
        }
    }

    // 3) Bonus for matching more tokens
    if (hits > 0 && hits === tokens.length) {
        score += 5;
    }

    // 4) If user message is exactly one of the tokens (e.g. "Panettone")
    if (tokens.length > 1 && tokens.some(tok => tok === msgNorm)) {
        score += 8;
    }

    return score;
};

const titles = [
    "Smoothie Giallo Passion",
    "Smoothie Rosso Berry",
    "Smoothie Verde Boost"
];

const query = "smoothie berry";

console.log(`Query: "${query}"`);
titles.forEach(title => {
    console.log(`Title: "${title}"`);
    const score = scoreTitle(title, query);
    console.log(`Score: ${score}`);
    console.log('---');
});
